### PR TITLE
Updated nextflow_schema.json

### PR DESCRIPTION
**Pull request by tmuylder**
Methylseq: Convert parameter docs to JSON schema. Based on the following issue: [#189](https://github.com/nf-core/methylseq/issues/189).

Added nextflow_schema.json file for methylseq. However, as the file already existed (in `dev` branch), it might be confusing to review. Someone with (more) experience in this specific pipeline, might have to evaluate whether the explanations and grouping are sufficient.

### CODE REVIEW

**Review by jfy133 — CHANGES_REQUESTED** [review]
First pass review - most of my current comments are for completeness more than anything. Can properly read through the messages next time :+1: 

I tried loading the schema in `nf-core schema build` and already noted a few things:

```
❓ Unrecognised 'params.reads' found in the schema but not in the pipeline config! Remove it? [y/n]:
✨ Found 'params.input' in the pipeline config, but not in the schema. Add to pipeline schema? [y/n]:
```

I'm assuming this is referring to the main input FASTQ files. Note that the 'new' nf-core standard is now to use `--input` instead of `--reads`, but this should be checked with the primary dev that using `--input` is OK.

There appears to be a range of missing params, so these also need to be added.

```
✨ Found 'params.em_seq' in the pipeline config, but not in the schema. Add to pipeline schema? [y/n]: y
✨ Found 'params.minins' in the pipeline config, but not in the schema. Add to pipeline schema? [y/n]: y
✨ Found 'params.maxins' in the pipeline config, but not in the schema. Add to pipeline schema? [y/n]: y
✨ Found 'params.publish_dir_mode' in the pipeline config, but not in the schema. Add to pipeline schema?
[y/n]: y
✨ Found 'params.name' in the pipeline config, but not in the schema. Add to pipeline schema? [y/n]: y
✨ Found 'params.project' in the pipeline config, but not in the schema. Add to pipeline schema? [y/n]: y

```

There are a lot of parameters without `--help` messages, but _with_ documentation.  Information. Try to summarise long documentation info into one sentence if possible for the `-help`, otherwise it's better to have _just_ help and no documentation if it is truly that simple a parameter [even though I see the `--help` message in dev doesn't have a help message at the moment]. this is for e.g. `--pbat`, `--single_cell`, `--epignome` etc.

Otherwise param types and also ordering etc. look good!

File: nextflow_schema.json
Comment (phue) [inline nextflow_schema.json:14]:
```suggestion
                "input": {
```

File: nextflow_schema.json
Comment (phue) [inline nextflow_schema.json:481]:
```suggestion
            "help_text": "Add arbitrary options for job submission.",
```

I think we can remove the profile specific hint here. This works for every executor that has the `clusterOptions` config field,

File: nextflow_schema.json
Comment (phue) [reply nextflow_schema.json:14]: I remember it used to be `--reads` but it was changed to `--input` in a previous template sync in an effort to standardize parameter names across pipelines.

File: nextflow_schema.json
Comment (phue) [inline nextflow_schema.json:48]:
```suggestion
                "input"
```
See L14

File: nextflow_schema.json
Comment (phue) [inline nextflow_schema.json:202]:
```suggestion
                    "description": "Skip deduplication step after alignment. This is turned on automatically if ` --rrbs` is specified.",
```

File: nextflow_schema.json
Comment (phue) [inline nextflow_schema.json:255]:
```suggestion
            "description": "Related to the tools used in nf-core/methylseq (Bismark, BWA-Meth and MethylDackel)",
```

File: nextflow_schema.json
Comment (phue) [inline nextflow_schema.json:54]:
```suggestion
            "description": "Depending on how they are prepared, bisulfite libraries often require additional base pairs to be removed from the ends of the reads before alignment. You can specify custom trimming parameters or use a preset as follows:",
```

**Review by phue — CHANGES_REQUESTED** [review]
Thanks for this, nice work!
I think this will really improve the methylseq documentation.
I added a few minor comments below :+1:

File: nextflow_schema.json
Comment (phue) [reply nextflow_schema.json:481]: @tmuylder 
Also, would you mind changing the all occurences of `clusterOptions` to `cluster_options`?
We changed the parameter naming to `snake_case` but apparently missed this one.

**Comment by apeltzer** [comment]
Guess @tmuylder this is already reviewed more than properly - shout if you need any help / comments 👍🏻

File: nextflow_schema.json
Comment (ewels) [inline nextflow_schema.json:14]: I don't think that this should be removed - input is required.

File: nextflow_schema.json
Comment (ewels) [inline nextflow_schema.json:16]: This is no longer the default

**Comment by ewels** [comment]
Please don't merge yet, I would like to have a proper review when I get back to my computer. I spotted a few problems already on my phone but tricky to see the diff when everything is moved around so will come back to this if that's ok.

**Review by ewels — CHANGES_REQUESTED** [review]
Will look through properly ASAP

File: nextflow_schema.json
Comment (tmuylder) [inline nextflow_schema.json:477]:
```suggestion
        "cluster_options": {
```

File: nextflow_schema.json
Comment (tmuylder) [inline nextflow_schema.json:479]:
```suggestion
            "description": "Arguments passed to Nextflow cluster_options.",
```

File: nextflow_schema.json
Comment (tmuylder) [reply nextflow_schema.json:14]: It's really confusing but the input is defined a bit lower (if you search for "required" you'll find it). All of the parameters in the whole doc have been moved up and down... I could have done my first-commit more efficiently...

File: nextflow_schema.json
Comment (tmuylder) [reply nextflow_schema.json:16]: I believe it is then best to remove a default there? 
It's still in the `nextflow.config` though. https://github.com/nf-core/methylseq/blob/master/nextflow.config

File: nextflow_schema.json
Comment (phue) [reply nextflow_schema.json:479]: In the description, it should remain `clusterOptions` because it is a [nextflow directive](https://www.nextflow.io/docs/latest/process.html#process-clusteroptions)
It's a bit confusing that the pipeline parameter is named the same, I know.

It needs to be changed to snake_case in the `nextflow.config` and the docs though

File: nextflow_schema.json
Comment (ewels) [reply nextflow_schema.json:16]: Ah right, it's about to not be: https://github.com/nf-core/methylseq/pull/194/files#diff-ce7465a8e67b3b9c95696db1146ca7e6328f075e08a3d64062b87aa8608fc4eaL13-R13 (also, stop looking at `master` 😉 )

Yeah the problem with having a default is that it's not really compatible with a parameter being `required` - if nothing is entered it uses the default and passes validation. So need to remove the default from the schema and set to `null` in the config.

File: nextflow_schema.json
Comment (ewels) [reply nextflow_schema.json:14]: Right, ok makes sense sorry - yes I had wondered which is why I gave up on my phone 😅

File: nextflow_schema.json
Comment (ewels) [reply nextflow_schema.json:479]: I will add this to `nf-core lint` so that `clusterOptions` is not allowed 👍🏻

Comment (apeltzer) [reply nextflow_schema.json:515]: How far are we here? I'm a bit lost on what needs to be done now

Comment (phue) [reply nextflow_schema.json:515]: @apeltzer I think this is good to go once the switch from `params.clusterOptions` -> `params.cluster_options` is made.
I would like to get this in before the 1.6 release, mostly because the Schema validation prints a warning if the parameter name is `CamelCase`

@tmuylder Would be great if you could find some time to :taco: this up, so we can have a release later :+1:

**Comment by ewels** [comment]
Hi all,

So if understand the background here @tmuylder - you didn't realise that we work on `dev` so made a schema from scratch. But we already had one on `dev` which I wrote a while back, so then you are trying to merge the schema you wrote onto dev onto the one I had...

I just tested `nf-core launch .` with the web interface on the existing `dev` code and the code in this PR and I hate to say it but I think that the existing schema is better 👀 As an end user I think that the groups we currently have make more sense. I clicked around a bit and most of the descriptions seemed the same etc.

Is there any proposed additional content here that I'm missing? Otherwise I'm wondering if we should close this PR and leave the schema as it is? Would be good if others could try both versions out as well and see what they think.

Sorry @tmuylder, I hope you don't take this personally at all - I realise that you were stung here by an accidental slip that is easy to make 😞 

Phil

**Comment by tmuylder** [comment]
he @ewels , no offense at all! I raised a PR because there was a comment that I should do it nonetheless it already existed (and was generated in an automatic way and hence not validated). I am more than happy to leave it as it is now (in the dev branch). So go ahead and close this PR, it's a learning process and now I know that I should ALWAYS start from the `dev` branch 😄

File: nextflow_schema.json
Comment (ewels) [inline nextflow_schema.json:100]: This is not the case - the original help text is correct here. TrimGalore doesn't have such a parameter. This is used in the following code:

https://github.com/nf-core/methylseq/blob/15cc666feb0b4085c30108a10e7ab120b9907ac8/main.nf#L125-L130

File: nextflow_schema.json
Comment (ewels) [inline nextflow_schema.json:105]: As with `--accel`

File: nextflow_schema.json
Comment (ewels) [inline nextflow_schema.json:110]: Also not correct.

**Review by ewels — CHANGES_REQUESTED** [review]
Found some more that is incorrect sorry. Ok to close without merging?

**Comment by ewels** [comment]
Thanks @tmuylder! Very gracious 🙇🏻 

It's obviously a bit late now, but we have some bytesize intro talks and stuff that go over our dev workflow 👍🏻

File: nextflow_schema.json
Comment (ewels) [reply nextflow_schema.json:515]: Looking at this again - the problem with this parameter is that it's a bit of a catch-22 situation.. It's used extensively in the centralised configs: https://github.com/nf-core/configs/search?q=clusteroptions

So updating in pipelines will mean that they need updating. And updating them will break them for any pipelines that haven't updated yet.. I think this is why we haven't done it yet.

But if linting is failing then yes that may be good enough motivation to do it. But should probably go out in a template sync to try to get all pipelines to change at once.

**Comment by phue** [comment]
> I raised a PR because there was a comment that I should do it nonetheless it already existed (and was generated in an automatic way and hence not validated).

Yes sorry about that, I might have confused the `nextflow_schema.json` with the `parameters.settings.json` that was there before (and was full of problems because I created that manually at some point)

**Pull request closed without merging**
